### DISCUSSION TITLE
RequestBody VS ModelAttribute ?

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#117] [Spring MVC]
.

**Comment by nayonsoso** [comment]
이들은 클라이언트 측에서 보낸 데이터를 Java 객체로 만들어주는데
RequestBody 는 `요청의 본문(Body)`에 있는 값을 바인딩할 때 사용하고,
ModelAttribute 는 `요청 파라미터나 multipart/form-data 형식`을 바인딩할 때 사용한다.

### RequestBody
- 클라이언트가 보내는 요청의 본문을 자바 객체로 변환한다.
- 내부적으로 HttpMessageConverter를 거치는데, 이때 ObjectMapper를 통해 JSON 값을 java 객체로 역직렬화한다. 
- 따라서 변환될 java 객체에 기본 생성자를 정의해야 하고, getter나 setter를 선언해야 한다. [참고](https://jenkov.com/tutorials/java-json/jackson-objectmapper.html#how-jackson-objectmapper-matches-json-fields-to-java-fields)
- cf. record에 기본 생성자를 따로 정의하지 않았는데 역직렬화가 되는 이유
 - record 는 기본생성자를 자동으로 제공하지 않는 대신, ’모든 필드를 초기화하는 생성자’를 제공한다.
  - jackson 은 일반 객체와 달리, record를 역직렬화할 때는 ’모든 필드를 초기화하는 생성자’를 사용해 역직렬화하기 때문이다.

### ModelAttribute
- 두가지 사용법이 있다.
- 첫번째 사용법인 메서드 단에서의 사용법은 jsp의 Model에 하나 이상의 속성을 추가하고 싶을 때 사용한다.
  - e.g. `model.addAttribute(“속성 이름”, “속성 값”)`
  - 이 방법의 내부 동작은 설명 생략하겠다.
- 두번째 사용법인 인자 단에서의 사용으로 클라이언트가 보내는 요청의 파라미터나 multipart/form-data 형식의 데이터를 자바 객체로 변환한다.
- 내부적으로 ModelAttributeMethodProcessor를 거치는데, 이때 지정된 클래스의 생성자를 찾아 객체로 변환한다.

ref. https://tecoble.techcourse.co.kr/post/2021-05-11-requestbody-modelattribute/
ref. https://blog.karsei.pe.kr/59

**Comment by gyungchan-jo** [comment]
### RequestBody

`@RequestBody` 애너테이션의 역할은 클라이언트가 보내는 HTTP 요청 본문(JSON 및 XML 등)을 Java 오브젝트로 변환하는 것입니다. HTTP 요청 본문 데이터는 Spring에서 제공하는 `HttpMessageConverter`를 통해 타입에 맞는 객체로 변환됩니다.

- `@RequestBody`를 사용하면 요청 본문의 JSON, XML, Text 등의 데이터가 적합한 HttpMessageConverter를 통해 파싱되어 Java 객체로 변환 된다.
- `@RequestBody`를 사용할 객체는 필드를 바인딩할 생성자나 setter 메서드가 필요없다.
- 직렬화를 위해 기본 생성자는 필수다.
- 데이터 바인딩을 위한 필드명을 알아내기 위해 getter나 setter 중 1가지는 정의되어 있어야 한다.

---

### ModelAttribute

`@ModelAttribute` 애너테이션의 역할은 클라이언트가 보내는 HTTP 파라미터들을 특정 Java Object에 바인딩(맵핑) 하는 것입니다. Query String 형태 혹은 요청 본문에 삽입되는 Form 형태의 데이터를 처리합니다.

- `@ModelAttribute`를 사용하면 HTTP 파라미터 데이터를 Java 객체에 맵핑한다.
- 따라서 객체의 필드에 접근해 데이터를 바인딩할 수 있는 생성자 혹은 setter 메서드가 필요하다.
- Query String 및 Form 형식이 아닌 데이터는 처리할 수 없다.

[참고 블로그](https://tecoble.techcourse.co.kr/post/2021-05-11-requestbody-modelattribute/)
